xds/src/main/java/io/grpc/xds/Filter.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.Any;
+import io.grpc.ClientInterceptor;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.ServerInterceptor;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.Nullable;
+
+/**
+ * Defines the parsing functionality of an HTTP filter. A Filter may optionally implement either
+ * {@link ClientInterceptorBuilder} or {@link ServerInterceptorBuilder} or both, indicating it is
+ * capable of working on the client side or server side or both, respectively.
+ */
+interface Filter {
+
+  /**
+   * The proto message types supported by this filter. A filter will be registered by each of its
+   * supported message types.
+   */
+  String[] typeUrls();
+
+  /**
+   * Parses the top-level filter config from raw proto message.
+   */
+  StructOrError<? extends FilterConfig> parseFilterConfig(Any rawProtoMessage);
+
+  /** Parses the per-filter override filter config from raw proto message. */
+  StructOrError<? extends FilterConfig> parseFilterConfigOverride(Any rawProtoMessage);
```

Comment:
This interface is redundant, you could just combine the two methods with
```
StructOrError<? extends FilterConfig> parseFilterConfig(Any rawProtoMessage, boolean isOverride);
```

Reply:
That's an anti-pattern. 

[Clean code: The curse of a boolean parameter](https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3)

[Java Clean Code Tutorial #4 - Avoid Boolean Parameters](https://www.youtube.com/watch?v=fiiEr0h9ChQ)

Reply:
In general `parseFilterConfig` and `parseFilterConfigOverride` can have completely different parsing logic, because the override config can be of a completely different type. It's only coincidence that the type of top-level config and override config are the same for Fault Filter.

Both [c-core](https://github.com/markdroth/grpc/blob/2d3cd132855e0421e2c2bf32881c4f11df249d85/src/core/ext/xds/xds_http_filters.h#L82) and [go](https://github.com/dfawley/grpc-go/blob/5a3bcfe0ac4e64e2ae874786a641b9b51fbc4bb9/xds/internal/httpfilter/httpfilter.go#L52) are using two methods rather than one single method with a boolean argument. It's natural to have two methods for the API and weird to have boolean argument.